### PR TITLE
[ Dy2Static ] transfer list into tensor array at runtime.

### PR DESCRIPTION
**Pull request by 2742195759**
### PR types

Others 
### PR changes

Others 
### Describe

1. make list transformer into jit form.
2. fix some bugs in tensor_array, such as append.
3. enhance the function analysis visitor to recognize push/pop.
4. add setter/getter helper to deal with 2+ name sets.

TODO:
1. remove the list_transformer file and other useless code.
2. support return value of partial_program layer is tensor_array:
```
def func(x):
    a = []
    for i in range(x):
         a.append(1)
    return a
```

### CODE REVIEW

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_operators.py [hunk 1]
```diff
@@ -91,21 +97,41 @@ def convert_while_loop(cond, body, getter, setter):
...
 
-def _run_paddle_while(cond, body, getter, setter):
+def _convert_tensor_arrray_if_necessary(setterhelper, push_pop_names):
+    push_pop_vars = setterhelper.get(push_pop_names)
+    if push_pop_vars is None:
+        return
+
+    def maybe_to_tensor_array(v):
+        if isinstance(v, list):
+            return create_array("float32", initialized_list=v)
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_operators.py:113]: 这里默认是float32类型，会不会有风险？比如是否会有其他数据类型的？

File: python/paddle/fluid/dygraph/dygraph_to_static/utils.py [hunk 2]
```diff
@@ -1315,3 +1347,57 @@ def create_nonlocal_stmt_nodes(names):
...
     func_code = "nonlocal {}".format(','.join(names))
     return [gast.parse(func_code).body[0]]
+
+
+class GetterSetterHelper:
+    """ we have two classes of names in setter and getter function: 
+        w_vars(loop_vars) + push_pop_vars
+        To simplify the setter logic in convert_while and convert_cond,
+        we extract the helper class here.
+    """
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/utils.py:1357]: Not Important, 后续注意这里的注释格式

File: python/paddle/fluid/layers/math_op_patch.py [hunk 3]
```diff
@@ -211,16 +212,35 @@ def append(self, var):
         
         """
         if not isinstance(var, Variable):
-            raise TypeError(
-                "Required input var should be Variable, but received {}".format(
-                    type(var)))
+            if in_declarative_mode():
+                """ in dy2static mode, x may be tensorable values such as int, float, np.array
+                """
```
Comment (Aurelius84) [inline python/paddle/fluid/layers/math_op_patch.py:217]: 单行注释可以用 #开头

**Review by Aurelius84 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_operators.py [hunk 4]
```diff
@@ -274,14 +306,15 @@ def convert_ifelse(pred, true_fn, false_fn, get_args, set_args,
         false_fn(callable): A callable to be performed if ``pred`` is false.
         get_args(callable): Get all arguments that needed in true_fn and false_fn.
         set_args(callable): Update arguments that modified in trure_fn and false_fn.
+        return_name_ids(list[string]): the returned names.
```
Comment (Ligoml) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_operators.py:309]: push_pop_names 是不是漏掉了？

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_operators.py [hunk 5]
```diff
@@ -74,7 +75,12 @@ def _unpack_by_structure_paddle(target, structure):
...
 
-def convert_while_loop(cond, body, getter, setter):
+def convert_while_loop(cond,
+                       body,
+                       getter,
+                       setter,
+                       return_name_ids=None,
+                       push_pop_names=None):
     """
     A function representation of a Python ``while`` statement.
```
Comment (Ligoml) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_operators.py:85]: 参数和文档中不一致，建议对照实际参数改一下哈

**Review by Ligoml — APPROVED** [review]
注释说明在 https://github.com/PaddlePaddle/Paddle/pull/44983 做统一修改吧

**Pull request merged**